scipy/optimize/_differentiable_functions.py

```diff
@@ -146,14 +146,14 @@ def update_x(x):
                 self.x_prev = self.x
                 self.g_prev = self.g
 
-                self.x = x
+                self.x = np.atleast_1d(x).astype(float)
```

Comment:
Why exactly are you using np.atleast_1d. Wouldn't it be better to use 'copy'.

I have the impression that  np.atleast_1d does not make a new copy all the time (see second example in [here](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.atleast_1d.html)), wouldn't that result in different behaviours for certain types of input... Making things even messier

Reply:
I started off my using `np.copy`. However, I thought it would be better to use the same behaviour as `__init__`, where the [same operation](https://github.com/scipy/scipy/blob/master/scipy/optimize/_differentiable_functions.py#L47) is performed. 
`np.atleast_1d` makes sure it's an array (but doesn't necessarily do a copy). `ndarray.astype(float)` does a ["Copy of the array, cast to a specified type"](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.astype.html).

So the behaviour in the PR is to make sure that the `x` attribute is definitely a float array, and a copy. 
The previous behaviour was to construct the `ScalarFunction` object with `x` as a copied float array, but on subsequent calls to `ScalarFunction.fun` the `x` parameter would become whatever was supplied, e.g. list of int, tuple of float16, etc.

I'm aware that dtypes supplied to construct the various `Function` objects are cast to np.float64. For simple function evaluation that may not be desired, but with numerical differentiation in `grad` and `hess` I think it's important that at least double precision be used.

Reply:
> I started off my using np.copy. However, I thought it would be better to use the same behaviour as __init__, where the same operation is performed.
np.atleast_1d makes sure it's an array (but doesn't necessarily do a copy). ndarray.astype(float) does a "Copy of the array, cast to a specified type".

Ok. Great that you have already considered this. I agree that is good to have the same behavior as init...